cmd/state-svc/internal/deprecation/deprecation.go

```diff
@@ -0,0 +1,200 @@
+package deprecation
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ActiveState/cli/internal/constants"
+	constvers "github.com/ActiveState/cli/internal/constants/version"
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/graph"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/multilog"
+	"github.com/ActiveState/cli/internal/retryhttp"
+	"github.com/hashicorp/go-version"
+	"github.com/patrickmn/go-cache"
+)
+
+const (
+	cacheKey = "info"
+)
+
+type ErrTimeout struct {
+	error
+}
+
+// info details deprecation information for a given version
+type info struct {
+	Version     string `json:"version"`
+	versionInfo *version.Version
+	Date        time.Time `json:"date"`
+	DateReached bool
+	Reason      string `json:"reason"`
+}
+
+// Checker is the struct that we use to do checks with
+type Checker struct {
+	config configurable
+	cache  *cache.Cache
+	done   chan struct{}
+}
+
+// configurable defines the configuration function used by the functions in this package
+type configurable interface {
+	ConfigPath() string
+	GetTime(key string) time.Time
+	Set(key string, value interface{}) error
+	Close() error
+}
+
+// NewChecker returns a new instance of the Checker struct
+func NewChecker(configuration configurable) *Checker {
+	checker := &Checker{
+		configuration,
+		cache.New(15*time.Minute, 5*time.Minute),
+		make(chan struct{}),
+	}
+
+	go checker.pollDeprecationInfo()
+
+	return checker
+}
+
+// Check will Check if the current version of the tool is deprecated and returns deprecation info if it is.
+// This uses a fairly short timeout to Check against our deprecation url, so this should not be considered conclusive.
+func (checker *Checker) Check() (*graph.DeprecationInfo, error) {
+	data, exists := checker.cache.Get(cacheKey)
+	if !exists {
+		return nil, nil
+	}
+
+	if !constvers.NumberIsProduction(constants.Version) {
+		return nil, nil
+	}
+
+	infos, ok := data.([]info)
+	if !ok {
+		return nil, errs.New("Unexpected cache entry for deprecation info")
+	}
+
+	versionInfo, err := version.NewVersion(constants.Version)
+	if err != nil {
+		return nil, errs.Wrap(err, "Invalid version number: %s", constants.Version)
+	}
+
+	for _, info := range infos {
+		if versionInfo.LessThan(info.versionInfo) || versionInfo.Equal(info.versionInfo) {
+			return &graph.DeprecationInfo{
+				Version:     info.Version,
+				Date:        info.Date.Format(constants.DateFormatUser),
+				DateReached: info.DateReached,
+				Reason:      info.Reason,
+			}, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (checker *Checker) pollDeprecationInfo() {
+	timer := time.NewTicker(1 * time.Minute)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			err := checker.Refresh()
+			if err != nil {
+				multilog.Critical("Could not update deprecation information %s", errs.JoinMessage(err))
```

Comment:
```suggestion
				multilog.Error("Could not update deprecation information %s", errs.JoinMessage(err))
```

Critical is for user facing non-input errors.

Reply:
`Critical` doesn't really communicate "user facing non-input errors" I would think that not being able to refresh the deprecation info would be fairly critical.

Reply:
Critical tells us on rollbar that the end-user was directly affected by the error. It's not meant to convey "this error is more serious than that error", it's just meant to convey "this error directly affected the user".

Reply:
What does "directly affect the user" mean here? The vast majority of our errors bubble up to the user but those aren't considered critical. By using critical for user-facing non-input errors we are saying that they are more serious than other errors. I'm being nit-picky here but I still think using critical logging for critical functionality makes sense. If deprecation is breaking it is something that should be conveyed to us in a more serious way. My misunderstanding could just be from my lack of familiarity with logging levels and what they are traditionally used for.